File: test_euler.py
#!/usr/bin/env python3

import subprocess
import sys

def test_euler4():
    """Test Euler4 implementation"""
    result = subprocess.run([sys.executable, 'Euler4.py'], capture_output=True, text=True)
    output = result.stdout.strip()
    expected = "906609"
    
    if expected in output:
        print(f"✓ Euler4 test passed: found {expected}")
        return True
    else:
        print(f"✗ Euler4 test failed: expected {expected}, got output: {output}")
        return False

def test_euler27():
    """Test Euler27 implementation"""
    result = subprocess.run([sys.executable, 'Euler27.py'], capture_output=True, text=True)
    output = result.stdout.strip()
    expected = "-59231"
    
    if expected in output:
        print(f"✓ Euler27 test passed: found {expected}")
        return True
    else:
        print(f"✗ Euler27 test failed: expected {expected}, got output: {output}")
        return False

if __name__ == "__main__":
    success1 = test_euler4()
    success2 = test_euler27()
    
    if success1 and success2:
        print("All tests passed!")
        sys.exit(0)
    else:
        print("Some tests failed!")
        sys.exit(1)

File: src/test_euler.hs
module Main where

import Euler27
import Euler4
import System.Exit (exitFailure, exitSuccess)

testEuler4 :: IO Bool
testEuler4 = do
    putStrLn "Testing Euler4 implementations..."

    let expected = 906609
    let results =
            [ largestPalindromeTailRec
            , largestPalindromeRec
            , largestPalindromeModular
            , largestPalindromeMap
            , largestPalindromeList
            ]

    let allCorrect = all (== expected) results

    if allCorrect
        then putStrLn ("All Euler4 implementations return " ++ show expected) >> return True
        else do
            putStrLn "Euler4 implementations disagree:"
            mapM_
                (\(name, result) -> putStrLn $ "  " ++ name ++ ": " ++ show result)
                [ ("TailRec", largestPalindromeTailRec)
                , ("Rec", largestPalindromeRec)
                , ("Modular", largestPalindromeModular)
                , ("Map", largestPalindromeMap)
                , ("List", largestPalindromeList)
                ]
            return False

testEuler27 :: IO Bool
testEuler27 = do
    putStrLn "Testing Euler27 implementations..."

    let expected = -59231
    let results =
            [ solutionTailRec
            , solutionRec
            , solutionModular
            , solutionMap
            , solutionList
            ]

    let allCorrect = all (== expected) results

    if allCorrect
        then putStrLn ("All Euler27 implementations return " ++ show expected) >> return True
        else do
            putStrLn "Euler27 implementations disagree:"
            mapM_
                (\(name, result) -> putStrLn $ "  " ++ name ++ ": " ++ show result)
                [ ("TailRec", solutionTailRec)
                , ("Rec", solutionRec)
                , ("Modular", solutionModular)
                , ("Map", solutionMap)
                , ("List", solutionList)
                ]
            return False

main :: IO ()
main = do
    success1 <- testEuler4
    putStrLn ""
    success2 <- testEuler27

    if success1 && success2
        then putStrLn "All tests passed!" >> exitSuccess
        else putStrLn "Some tests failed!" >> exitFailure